### DISCUSSION TITLE
Scheduled backups

### DISCUSSION THREAD

**Discussion by adamburgoyne** [mealie-recipes/mealie#4223] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/mealie-recipes/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

Currently there doesn't appear to be a way of running a scheduled backup. You can run one from the UI but that's a disaster waiting to happen.

### Please provide a concise description of the feature that would resolve your issue.

We all know the value of backups but how often do you manually create one... never. The ability to schedule backups which can then automatically be uploaded to a cloud provider would tempt me to move over my whole recipe collection.

### Please consider and list out some caveats or tradeoffs made in your design decision

There are none. 

### Additional Information

- [ ] If this is accepted I'm willing to submit a PR to provide this feature
- [ ] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by Kuchenpirat** [comment]
we always had the stance that it is best to use an external tool for backup functions. 
Everything you need is in the mounted folder if you are using sqlite. If you are using postgres as database you have to backup that too.

If you do not want to use an external tool you could also use a script to create backups via the api which should be possible to be easily automated.

**Reply by adamburgoyne** [reply]
Whilst I don't share the same stance I appreciate that it's been thought about! 

I'll take a look at calling the API to generate and pull backups. I think I'd prefer to trigger the backup process and upload the zip file to my NAS, that way if there's any db corruption that I don't spot for a while I'm not just backing up a corrupted db. Makes it a bit easier to pull an old version of mealie and just upload the file if something does go wrong at any point. Should be easy enough to knock up a bash script for!

**Reply by zolakt** [reply]
@Kuchenpirat Don't take this the wrong way, but I'm not really following the logic here. Basically you advise to use external tools for backups, I assume something like restic, borg, pbs etc. And that is fine, however, if that is the case, why have an internal backup/restore mechanism at all? Since you do have it, it would really be convient if it had a schedule and retention policy, e.g. similar to how immich, home assistant or vaultwarden-backup has it. It doesn't contradict your recommended external backup tool approach, it just makes it more flexible.

This is purely personal preference, and maybe I'm doing it wrong, but I like to keep system backups and service db backups separate. I plan to run this in a LXC container. I backup all containers with PBS, and that will work fine. However, I host multiple services in each LXC, and just because I mess up some recipes (nothing system related), I don't want to revert other services to older versions. Sure, I could restore a full backup somewhere else, cherrypick your folder and copy-paste it back manually... but that's a hastle. Considering you already have an internal backup/restore mechanism, it would be much more convenient to just restore a db backup from inside mealie, and not mess with PBS at all in this case.

**Reply by Kuchenpirat** [reply]
Long story short, we used to have this in mealie, it was kinda complex to make fit the different backup strategies of people so we have decided to defer backup strategies to the server admin. 
You could build some automation pretty easily over the api, that creates a backup every night and deletes backups that are older than x days. But i think some autorestic jobs are easier to handle.

The other advantage of external backups is that you are not relying on mealie which would make you beholden to any bugs we may have within the backup restore so you could be SOL if we find some game breaking bug down the line. Will that happen? Hard to say, but for data preservation purposes, I always suggest backing up directly.

**Reply by zolakt** [reply]
Yeah, I know I can script it. It just seemed a little strange that you advocate external backups (which is a valid stance), but then still have an internal backup/restore mechanism.

Just one question regarding scripting. Are the backups triggered from UI/API tracked in the db in any way? In other words, is it safe to just delete the zip files from the filesystem, without having some orphaned mess left behind?

**Reply by Kuchenpirat** [reply]
For once, it comes in quite handy as a dev when hadling multiple branches with different migrations applied and it is something basic to have build in for users that don't have an external backup solution. 

I don't say i don't see the values in an all included solution. But for now it is not included in mealie. 

I think you should be fine just deleting backups. 
If i remmeber right, backups are not stored in the database. Am just on phone currently, so cant verify right now.

**Reply by zolakt** [reply]
Thanks for the explanation. It clears up everything. Just please verify it's safe to just delete the zip files, so I can do some poor-mans retention policy in the script.

In the end, I'm actually considering keeping both ways of backup. I doubt it will take much disk space. So I have the scripted ones in case I do some mess in mealie. And if all hell breaks loose, I still have a full, self suficient, PBS backup of the whole service

**Reply by Kuchenpirat** [reply]
you should be good to go just deleting backups. There should not be anything in the db

**Comment by adamburgoyne** [comment]
Very easy in the end if you want to do something similar but requires a little manual intervention. 

I added `- ${NAS_MOUNT}/mealie-backups:/app/data/backups` to my docker-compose file which maps the mealie backup folder directly to my NAS drive (which then has a delete policy on the backup folder).

I've then just scheduled a nightly cron job that calls a python script to send a POST request to the backup endpoint to create a new one, the response is checked and if successful HealthChecks.io is updated. 

All sorted!

**Reply by Kuchenpirat** [reply]
Thanks for reporting back and glad to hear you sorted something out that works for you 👍

**Reply by chuckfy** [reply]
> 
> I've then just scheduled a nightly cron job that calls a python script to send a POST request to the backup endpoint to create a new one, the response is checked and if successful HealthChecks.io is updated.
> 
> All sorted!

Would you mind sharing that script?

**Reply by adamburgoyne** [reply]
@chuckfy I don't really like python but it seemed like a good fit for the job so I just asked chatgpt to knock me up a quick script. It's pretty basic but it does the job. 

```
import requests

mealieBackupUrl = "http://ip:port/api/admin/backups"

payload = {}
headers = {
  'Accept-Language': 'application/json',
  'Authorization': 'Bearer blah'
}

try:
     # Send POST request
    response = requests.post(mealieBackupUrl, headers=headers, data=payload)

    # Check if response status code is 201 and error field is false
    if response.status_code == 201:
        response_json = response.json()
        if response_json.get("error") == False:
            print("Backup created successfully.")
            # Send the healthcheck ping
            requests.get("https://hc-ping.com/blah", timeout=10)
        else:
            print("Error in backup response: %s" % response_json.get("message"))
    else:
        print("Failed to create backup, status code: %d" % response.status_code)

except requests.RequestException as e:
    # Log ping failure here...
    print("Ping failed: %s" % e)
```

**Reply by chuckfy** [reply]
> @chuckfy I don't really like python but it seemed like a good fit for the job so I just asked chatgpt to knock me up a quick script. It's pretty basic but it does the job.

Greatly appreciated!

**Comment by ananthb** [comment]
I solved this with a shell script that calls the admin backups API before it backs up the `mealie/backups` folder.
